Terraform Import (#216)
earthly/website: blog/_posts/2022-07-12-terraform-lambda.md

---
title: "Terraform Import - Leaving Click Opps Behind"
categories:
  - Tutorials
toc: true
sidebar:
  nav: "lambdas"
author: Adam

internal-links:
 - terraform import
---
<!-- vale on -->
## From Click Ops To GitOps

Previously I built a [REST API](/blog/aws-lambda-api-proxy/), deployed it into a container and got it all running on AWS as a Lambda. But setting this up involved just clicking around in AWS and occasionally using the AWS CLI.

Today, I'm going to port that whole setup to Terraform so that its easier to manage, reproduce, and make changes to. ( These are the purported benefits of Terraform, or so my coworker Corey tells me. I've never used it before so I'll be learning as I go. )

> **Read this to learn about Terraform from a noob's perspective. The specifics of my infrastructure import may not be relevant to you, but hopefully my thinking process and take-aways will be.**

Author: yeah, makes sense. 

There are probably less explicit ways to do it, but I think of it as saying "Here is why you should  (or shouldn't) read this" somewhere right at the beginning.

~~~{.bash caption=">_"}
terraform plan
~~~

~~~{.ini caption="Output"}
╷
│ Error: error configuring Terraform AWS Provider: no valid credential sources for Terraform AWS Provider found.
~~~

If the credentials you have are wrong, the error will be a little bit different:

~~~{.ini caption=""}
Error: error configuring Terraform AWS Provider: error validating provider credentials: error calling sts:GetCallerIdentity: operation error STS: GetCallerIdentity, https response error StatusCode: 403, RequestID: e816a4d9-6f28-49ef-b245-395c3a758f4a, api error InvalidClientTokenId: The security token included in the request is invalid.
~~~

To debug this you need to understand that the aws provider reads from `~/.aws/confg` and `~/.aws/credentials`, the same way the `aws` cli does. If you have credentials for more than one AWS account available, you can use `AWS_PROFILE` to configure the AWS provider. In my case I do this like so:

~~~{.bash caption=">_"}
$ export AWS_PROFILE=earthly-dev
~~~

It's also possible to configure the profile in terraform by adding it under provider configuration:

> ~~~{.groovy caption="main.tf"}

Author: Somehow this wasn't happening once I made other changes. Hopefully it doesn't come back :)

~~~{.bash caption=">_"}
$ terraform apply --auto-approve
~~~

~~~{.ini caption=""}
Plan: 1 to add, 0 to change, 0 to destroy.
aws_ecr_repository.lambda-api: Creating...
╷
│ Error: failed creating ECR Repository (lambda-api): 
|  RepositoryAlreadyExistsException: The repository with name 'lambda-api'
|   already exists in the registry with id '459018586415'
│   with aws_ecr_repository.lambda-api,
│   on main.tf line 14, in resource "aws_ecr_repository" "lambda-api":
│   14: resource "aws_ecr_repository" "lambda-api" {
│ 
╵
~~~

You can see that Terraform calls the AWS API , which fails to create a new repository, because one already existed. This is going to be an ongoing problem because everything I'd like to get into terraform today I've already set up using the AWS UI.

There are a couple ways to manage this.

- Remove everything via UI and recreate in Terraform.
  
  I reached out to Corey, my local Terraform expert and this is the path he recommended:

  > Use the UI once to figure out how to set things up but then tear it down and recreate it in Terraform.  

- Use a 'reverse terraformer'.

>   Tools exist that proport to take an existing cloud accounts and import them into Terraform state and generated the Terraform config for them. I did briefly try this approach.

Reviewer: ```suggestion
  Tools exist that purport to take existing cloud accounts and import them into Terraform state and generate the Terraform config for them. I did briefly try this approach.
```

  Prepared aws_ecr_repository for import
aws_ecr_repository.lambda-api: Refreshing state... [id=lambda-api]

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.
~~~

Then I apply:

~~~{.bash caption=">_"}
terraform apply --auto-approve
~~~

~~~{.ini caption=""}
aws_ecr_repository.lambda-api: Refreshing state... [id=lambda-api]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and
found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
~~~

And I'm all set: My repository is now managed by Terraform.

## Learning From Terraform State with `terraform show`

The Terraform docs are often terse and values in the resource config do not always map one-to-one onto values you'd find the AWS UI. For me, this meant that I couldn't always figure out exactly what I needed to put in the resource configuration to match what I had set up in the UI. Thankfully you can use `terraform show` to retrieve this information.

> Let's pretend that I needed my repo to use KMS encryption and to scan image on push. I've already setup these options in AWS but I want to see how to do it AWS.

Reviewer: ```suggestion
Let's pretend that I needed my repo to use KMS encryption and to scan an image on push. I've already setup these options in AWS but I want to see how to do it in Terraform.
```

$ terraform show
~~~

~~~{.groovy caption="Output"}
# aws_ecr_repository.lambda-api:
resource "aws_ecr_repository" "lambda-api" {
    arn                  = "arn:aws:ecr:us-east-1:459018586415:repository/
                            lambda-api"
    id                   = "lambda-api"
    image_tag_mutability = "MUTABLE"
    name                 = "lambda-api"
    registry_id          = "459018586415"
    repository_url       = "459018586415.dkr.ecr.us-east-1.amazonaws.com/
                            lambda-api"
    tags                 = {}
    tags_all             = {}

    encryption_configuration {      # <- Here is what I need!
        encryption_type = "KMS"
    }

    image_scanning_configuration {  # <- Here also !
        scan_on_push = true
    }

    timeouts {}
}
~~~

And I can easily see how to set the encryption and image scanning options and therefor copy them into my `main.tf`.

~~~{.groovy caption="main.tf"}
resource "aws_ecr_repository" "lambda-api" {
  image_tag_mutability = "MUTABLE"
  name                 = "lambda-api"
  encryption_configuration {
      encryption_type = "KMS"
  }
  image_scanning_configuration {
        scan_on_push = true
  }
}
~~~

> This trick was instrumental in me getting my lambda integration ported over to terraform, but I'll refrain from focusing on it depth. Just remember, if you can't figure out how to configure something in Terraform for some reason, one approach is to set it up in the UI first, and use `terraform import` and `terraform show` to extract a working config.

Reviewer: ```suggestion
This trick was instrumental in me getting my lambda integration ported over to terraform, but I'll refrain from focusing on it in depth. Just remember, if you can't figure out how to configure something in Terraform for some reason, one approach is to set it up in the UI first, and use `terraform import` and `terraform show` to extract a working config.
```

## AWS ECR Policy Resource Import

Next up is the ECR Policy. First I created it as a blank resource:

~~~{.groovy caption="main.tf"}
resource "aws_ecr_repository_policy" "lambda-api" {
  }
~~~

> Imported is:

Reviewer: ```suggestion
Imported with:
```

~~~{.bash caption=">_"}
$ terraform import aws_ecr_repository_policy.lambda-api lambda-api
~~~

> And then with the show trick I extracted the policy I was already using:

Reviewer: ```suggestion
And then, with the `terraform show` trick I used above, I extracted the policy I was already using:
```

~~~

Then I run apply, which will find no changes to apply.

You may be wondering how certain I can be in all this declarative config, when none of it has actually been applied from terraform, only imported. Well I'm going to tackle that last.

## Terraform API Gateway Import

Next up is my API Gateway.

First I need to import my domain certificate and set up the API gateway to use it.

~~~{.groovy caption="main.tf"}
## API GATEWAY

# ## Domain Name

resource "aws_acm_certificate" "earthly-tools-com" {
    domain_name               = "earthly-tools.com"
    subject_alternative_names = [
        "earthly-tools.com",
    ]

    options {
        certificate_transparency_logging_preference = "ENABLED"
    }
}

resource "aws_api_gateway_domain_name" "earthly-tools-com" {
  domain_name              = "earthly-tools.com"
  regional_certificate_arn = aws_acm_certificate.earthly-tools-com.arn

  endpoint_configuration {
    types = ["REGIONAL"]
  }
}
~~~

And import it by ARN, which I got from the AWS UI:

~~~{.bash caption=">_"}
$ terraform import aws_acm_certificate.cert arn:...
$ terraform import aws_api_gateway_domain_name.earthly-tools-com earthly-tools.com
~~~

> Then I need to set up the API, the API State and map it to the domain.

Reviewer: Should this be API stage?

Author: Yes, nice catch.

  timeout                        = "120"
  architectures = ["x86_64"]

  environment {
    variables = {
      HOME = "/tmp"
    }
  }

  ephemeral_storage {
    size = "512"
  }

  tracing_config {
    mode = "PassThrough"
  }
}
~~~

I did a similar process for the lambda permissions, since the API needs permissions to call the lambda. The end result looks like this:

~~~{.groovy caption="main.tf"}
## Give API Gateway access to Lambda
resource "aws_lambda_permission" "earthly-tools-com" {
  action        = "lambda:InvokeFunction"
  function_name = aws_lambda_function.lambda-api.function_name
  principal     = "apigateway.amazonaws.com"
  source_arn    = "arn:aws:execute-api:us-east-1:459018586415:
                  ${aws_apigatewayv2_api.earthly-tools-com.id}/*/*/{path+}"
}
~~~

Note that `terraform show` returned the current api arn for `source_arn` and I needed to replace that with a path to make this future proof. This is something to keep vigilant for when using `terraform show` or you might end up with resources hardcoded to values that quickly go out of date.

## Terraform Import API Route and Integration

> Integrating the lambda with the API was by far the hardest part for me. In AWS, there is a single button click process to hook a lambda up to an API endpoint. Behind the scenes there are a number of separate things happening. But once I figured out the name of all these little resources, it was easy to import and 'terraform show` them to see how they should be configured.

Reviewer: ```suggestion
Integrating the lambda with the API was by far the hardest part for me. In AWS, there is a single button click process to hook a lambda up to an API endpoint. Behind the scenes there are a number of separate things happening. But once I figured out the name of all these little resources, it was easy to import and `terraform show` them to see how they should be configured.
```

aws_apigatewayv2_api.earthly-tools-com: Destruction complete after 1s

Apply complete! Resources: 0 added, 0 changed, 12 destroyed.
~~~

At which point, I can confirm nothing worked:

~~~{.bash caption=">_"}
curl https://earthly-tools.com/text-mode
{"message":"Internal Server Error"}%        
~~~

Then I started un-commenting resources working from the top down and applying as I went. Until I got this:

~~~{.bash caption=">_"}
aws_ecr_repository_policy.lambda-api: Creation complete after 1s [id=lambda-api]
╷
│ Error: error creating Lambda Function (1): InvalidParameterValueException: Source image 459018586415.dkr.ecr.us-east-1.amazonaws.com/lambda-api:latest does not exist. Provide a valid source image.
│ {
│   RespMetadata: {
│     StatusCode: 400,
│     RequestID: "9b458fc2-7897-4d46-88f0-5b52494ec276"
│   },
│   Message_: "Source image 459018586415.dkr.ecr.us-east-1.amazonaws.com/lambda-api:latest does not exist. Provide a valid source image.",
│   Type: "User"
│ }
│ 
│   with aws_lambda_function.lambda-api,
│   on main.tf line 126, in resource "aws_lambda_function" "lambda-api":
│  126: resource "aws_lambda_function" "lambda-api" {
│ 
╵
~~~

> In turns out that the lambda needs to have a container in place before being created. So I uncommented the lambda and did a docker push

Reviewer: ```suggestion
It turns out that the lambda needs to have a container in place before being created. So I uncommented the lambda and did a docker push
```
